# A2: Vector Semantics

Nikolai Ilinykh, Mehdi Ghanimifard, Wafia Adouane and Simon Dobnik

The lab is an exploration and learning exercise to be done in a group and also in discussion with the teachers and other students.

Before starting, please read [the following instructions](https://github.com/sdobnik/computational-semantics/blob/master/README.md) on how to work on group assignments.

Write all your answers and the code in the appropriate boxes below.

In this lab we will look at how to build distributional semantic models from corpora and use semantic similarity captured by these models to do semantic tasks. We are also going to examine how different vector composition functions for phrases affect both the model and the learned information about similarities.  

Note that this lab uses a code from `dist_erk.py`, which contains functions that highly resemble those shown during the lecture. In the end, you can use either of the functions (from the lecture / from the file) to solve the tasks.

In [1]:
# the following command simply imports all the methods from that code.
from dist_erk import *

## 1. Loading a corpus

To train a distributional model, we first need a sufficiently large collection of texts which contain different words used frequently enough in different contexts. Here we will use a section of the Wikipedia corpus which you can download from [here](https://linux.dobnik.net/cloud/index.php/s/isMBj49jt5renYt?path=%2Fresources%2Fa2-distributional-representations) (wikipedia.txt.zip). (This file has been borrowed from another lab by [Richard Johansson](http://www.cse.chalmers.se/~richajo/)).  
When unpacked, the file is 151mb, hence if you are using the MLT servers you should store it in a temporary folder outside your home and adjust the `corpus_dir` path below.  
<!-- <It may already exist in `/opt/mlt/courses/cl2015/a5`.> -->


In [2]:
corpus_dir = '/root/Ourania_Kolovou/02-vector-semantics/assignment/'

## 2. Building a model

Now you are ready to build the model.  
Using the methods from the code imported above build three word matrices with 1000 dimensions as follows:  

(i) with raw counts (saved to a variable `space_1k`);  
(ii) with PPMI (`ppmispace_1k`);  
(iii) with reduced dimensions SVD (`svdspace_1k`).  
For the latter use `svddim=5`. **[5 marks]**

Your task is to replace `...` with function calls. Functions are imported from `dist_erk.py` earlier, and they largely resemble functions shown during the lecture.

In [3]:
numdims = 1000
svddim = 5

# which words to use as targets and context words?
# we need to count the words and keep only the N most frequent ones
# which function would you use here with which variable?
ktw = do_word_count(corpus_dir, numdims)

wi = make_word_index(ktw) # word index
words_in_order = sorted(wi.keys(), key=lambda w:wi[w]) # sorted words

# create different spaces (the original matrix space, the ppmi space, the svd space)
# which functions with which arguments would you use here?
print('create count matrices')
space_1k = make_space(corpus_dir, wi, numdims)
print('ppmi transform')
ppmispace_1k = ppmi_transform(space_1k, wi)
print('svd transform')
svdspace_1k = svd_transform(ppmispace_1k, numdims, svddim)
print('done.')

reading file wikipedia.txt
create count matrices
reading file wikipedia.txt
ppmi transform
svd transform
done.


In [4]:
# now, to test the space, you can print vector representation for some words
print('house:', space_1k['house'])
print('air:', ppmispace_1k['air'])
print('sea:', svdspace_1k['sea'])

house: [2554 3774 3105  567  962  631  443  185  311  189  131   28   93  169
   81  125  151  408  194   90   79   29  217  184   62   15   31   70
   10    1   41   21    1   31   37    1   30    5   25    7    3   20
   11    1   32   36    2    5   66    4    0   46    8   18   28    0
   20    7    8   16   10   40    0  175   10    2    7   19    1  174
   11    3    1    6    0    0    0   10    9   11    7   24    4    4
   14   23   58    7    0   10    2    3   10    6   18    6   13    3
   22    0    3    5    3    7   14    3   40   20   19   15    6    8
   24    4    5    1   19    0    3    1    0   14    0   14   53    7
    7   11    6    5    5    4   12    6   53    1    1  433    4    0
    5    7    7   12    1    1    3    4   17    8   16    1    2   31
    1   12   14    1   44    6   14    9   38    7    2    6    8    1
   10    6   10    1    9    7    9    4    3   10    0   11    3    2
    0    2   11   37    2    0    2    1    5    9   10   16   88    6

Oxford Advanced Dictionary has 185,000 words, hence 1,000 words is not representative. We trained a model with 10,000 words, and 50 dimensions on truncated SVD. It took 40 minutes on a laptop. We saved all three matrices [here](https://linux.dobnik.net/cloud/index.php/s/isMBj49jt5renYt?path=%2Fresources%2Fa2-distributional-representations) (pretrained.zip). Download them and unpack them to a `pretrained` folder which should be a subfolder of the folder with this notebook:

In [5]:
import numpy as np

numdims = 10000
svddim = 50

print('Please wait...')
ktw_10k       = np.load('./pretrained/ktw_wikipediaktw.npy', allow_pickle=True)
space_10k     = np.load('./pretrained/raw_wikipediaktw.npy', allow_pickle=True).all()
ppmispace_10k = np.load('./pretrained/ppmi_wikipediaktw.npy', allow_pickle=True).all()
svdspace_10k  = np.load('./pretrained/svd50_wikipedia10k.npy', allow_pickle=True).all()
print('Done.')


Please wait...
Done.


In [6]:
# testing semantic space
print('house:', space_10k['house'])
print('air:', ppmispace_10k['air'])
print('sea:', svdspace_10k['sea'])

house: [2554 3774 3105 ...    0    0    0]
air: [0.         0.20245295 0.         ... 0.         0.         0.        ]
sea: [-13.67271457   0.43142529  -6.94486778  -2.07479418   2.31061203
  -8.37480138   3.38127483   1.38862392   3.69919853   4.83386322
   1.20483282  -2.85352271   2.42117201  -3.19388172  -2.40011153
  -0.20971372  -3.24308822   3.45394695   0.44410338  -0.28501801
  -1.01970338  -3.79003969  -1.57024237  -0.66787731   1.66208673
  -1.06802383   0.23360065   4.40315485   0.63124451   2.07411325
  -0.61039791  -3.45987008   1.50906565   0.68859961   1.56440767
   1.13997986  -4.09802398  -1.51193416  -2.41297892   1.33231128
   1.76428807   0.2758639    2.87647971  -3.26989282   0.43006076
  -3.26029135  -1.15181694   0.34613901  -5.99705027  -1.11413071]


## 3. Testing semantic similarity

The file `similarity_judgements.txt` (a copy is included with this notebook) contains 7,576 pairs of words and their lexical and visual similarities (based on the pictures) collected through crowd-sourcing using Mechanical Turk as described in [1]. The score range from 1 (highly dissimilar) to 5 (highly similar). Note: this is a different dataset from the phrase similarity dataset we discussed during the lecture (the one from [2]). For more information, please read the papers.

The following code will transform similarity scores into a Python-friendly format:

In [7]:
word_pairs = [] # test suit word pairs
semantic_similarity = [] 
visual_similarity = []
test_vocab = set()

for index, line in enumerate(open('similarity_judgements.txt')):
    data = line.strip().split('\t')
    if index > 0 and len(data) == 3:
        w1, w2 = tuple(data[0].split('#'))
        # it will check if both words from each pair exist in the word matrix.
        if w1 in ktw_10k and w2 in ktw_10k:
            word_pairs.append((w1, w2))
            test_vocab.update([w1, w2])
            semantic_similarity.append(float(data[1]))
            visual_similarity.append(float(data[2]))
        
print('number of available words to test:', len(test_vocab-(test_vocab-set(ktw))))
print('number of available word pairs to test:', len(word_pairs))
#list(zip(word_pairs, visual_similarity, semantic_similarity))

number of available words to test: 13
number of available word pairs to test: 774


Now we are going to test how the cosine similarity between vectors of each of the three spaces (normal space, ppmi, svd) compares with the human similarity judgements for the words in the similarity dataset. Which of the three spaces best approximates human judgements?

For comparison of several scores, we can use [Spearman correlation coefficient](https://en.wikipedia.org/wiki/Spearman's_rank_correlation_coefficient) which is implemented in `scipy.stats.spearmanr` [here](https://docs.scipy.org/doc/scipy-0.19.0/reference/generated/scipy.stats.spearmanr.html). The values of the Sperman correlation coefficient range from -1, 0 to 1, where 0 indicates no correlation, 1 perfect correaltion and -1 negative correlation. Hence, the greater the number the better the similarity scores align. The p values tells us if the coefficient is statistically significant. For this to be the case, it must be less than or equal to $< 0.05$.

Here is how you can calculate Pearson's correlation coefficient betweeen the scores of visual similarity and semantic similarity of the available words in the test suite:

In [8]:
from scipy import stats

rho, pval = stats.spearmanr(semantic_similarity, visual_similarity)
print("""Visual Similarity vs. Semantic Similarity:
rho     = {:.4f}
p-value = {:.4f}""".format(rho, pval))


Visual Similarity vs. Semantic Similarity:
rho     = 0.7122
p-value = 0.0000


Let's now calculate the cosine similarity scores of all word pairs in an ordered list using all three matrices. **[6 marks]**

In [9]:
raw_similarities  = [cosine(w1, w2, space_10k) for w1, w2 in word_pairs]
ppmi_similarities = [cosine(w1, w2, ppmispace_10k) for w1, w2 in word_pairs]
svd_similarities  = [cosine(w1, w2, svdspace_10k) for w1, w2 in word_pairs]

Now, calculate correlation coefficients between lists of similarity scores and the real semantic similarity scores from the experiment. The scores of what model best correlates them? Is this expected? **[6 marks]**

In [10]:
# your code should go here
rho, pval = stats.spearmanr(semantic_similarity, raw_similarities)
print("""Semantic_similarity vs Raw_Similarity
rho     = {:.4f}
p-value = {:.4f}""".format(rho, pval))
print()

rho1, pval1 = stats.spearmanr(semantic_similarity, ppmi_similarities)
print("""Semantic_Similarity vs Ppmi_Similarity:
rho     = {:.4f}
p-value = {:.4f}""".format(rho1, pval1))
print()

rho2, pval2 = stats.spearmanr(semantic_similarity, svd_similarities)
print("""Semantic_Similarity vs Svd_Similarity:
rho     = {:.4f}
p-value = {:.4f}""".format(rho2, pval2))
print()

Semantic_similarity vs Raw_Similarity
rho     = 0.1522
p-value = 0.0000

Semantic_Similarity vs Ppmi_Similarity:
rho     = 0.4547
p-value = 0.0000

Semantic_Similarity vs Svd_Similarity:
rho     = 0.4232
p-value = 0.0000



**Your answer should go here:**

The model with the best correlation is the PPMI. It is expected PPMI and SVD to have better results than Raw. PPMI has a slight difference on score than SVD with 0.4547 and 0.4232 respectively. 

We can also calculate correlation coefficients between lists of cosine similarity scores and the real visual similarity scores from the experiment. Which similarity model best correlates with them? How do the correlation coefficients compare with those from the previous comparison - and can you speculate why do we get such results? **[7 marks]**

In [11]:
# Your code should go here...
rho, pval = stats.spearmanr(visual_similarity, raw_similarities)
print("""Visual_similarity vs Raw_Similarity
rho     = {:.4f}
p-value = {:.4f}""".format(rho, pval))
print()

rho1, pval1 = stats.spearmanr(visual_similarity, ppmi_similarities)
print("""Visual_Similarity vs Ppmi_Similarity:
rho     = {:.4f}
p-value = {:.4f}""".format(rho1, pval1))
print()

rho2, pval2 = stats.spearmanr(visual_similarity, svd_similarities)
print("""Visual_Similarity vs Svd_Similarity:
rho     = {:.4f}
p-value = {:.4f}""".format(rho2, pval2))
print()

Visual_similarity vs Raw_Similarity
rho     = 0.1212
p-value = 0.0007

Visual_Similarity vs Ppmi_Similarity:
rho     = 0.3838
p-value = 0.0000

Visual_Similarity vs Svd_Similarity:
rho     = 0.3097
p-value = 0.0000



**Your answer should go here:**

In thsese results PPMI has again the best correlation that the other two models(RAW, SVD).
RAW model has the lowest score of af all, but SVD's score is close to the PPMI's. In general, scores of semantic similarity are higher than visual similarity for all models. Also, here we can see that raw similarity has a p-value score different than 0. In semantic similarity p-value was 0 for all models.   

## 4. Operations on similarities

We can perform mathematical operations on vectors to derive meaning predictions. For example, we can subtract the normalised vectors for `king` minus `queen` and add the resulting vector to `man` and we hope to get the vector for `woman`. Why? **[3 marks]**

**Your answer should go here:**

Here is some helpful code that allows us to calculate such comparisons.

In [12]:
from scipy.spatial import distance

def normalize(vec):
    return vec / veclen(vec)

def find_similar_to(vec1, space):
    # vector similarity funciton
    #sim_fn = lambda a, b: 1-distance.euclidean(normalize(a), normalize(b))
    #sim_fn = lambda a, b: 1-distance.correlation(a, b)
    #sim_fn = lambda a, b: 1-distance.cityblock(normalize(a), normalize(b))
    #sim_fn = lambda a, b: 1-distance.chebyshev(normalize(a), normalize(b))
    #sim_fn = lambda a, b: np.dot(normalize(a), normalize(b))
    sim_fn = lambda a, b: 1-distance.cosine(a, b)

    sims = [
        (word2, sim_fn(vec1, space[word2]))
        for word2 in space.keys()
    ]
    return sorted(sims, key = lambda p:p[1], reverse=True)

Here is how you apply this code. Comment on the results you get. **[3 marks]**

In [13]:
short = normalize(svdspace_10k['short'])
light = normalize(svdspace_10k['light'])
long = normalize(svdspace_10k['long'])
heavy = normalize(svdspace_10k['heavy'])

find_similar_to(light - (heavy - long), svdspace_10k)[:10]

[('long', 0.8733111261346901),
 ('above', 0.8259671977311955),
 ('around', 0.8030776291120685),
 ('sun', 0.7692439111243973),
 ('just', 0.7678481974778111),
 ('wide', 0.767257431992253),
 ('each', 0.7665960260861158),
 ('circle', 0.7647746702909336),
 ('length', 0.7601066921319761),
 ('almost', 0.7542351860536628)]

**Your answer should go here:**

I think it tries to find similar words with 'short', but there are some words which I do not know how could be similar. For example how sun, circle, almost ,just can be similar to short??


Find 5 similar pairs of pairs of words and test them. Hint: Google for `word analogies examples`. You can also construct analogies that are less lexical but more grammatical, e.g. `see, saw, leave, ?` or analogies that are based on world knowledge as in the [Google analogy dataset](http://download.tensorflow.org/data/questions-words.txt) from [3]. Does the resulting vector similarity confirm your expectations? But remember you can only do this if the words are contained in our vector space with 10,000 dimensions. **[10 marks]**

In [14]:
# Your code should go here...
#words similar to largest
large = normalize(svdspace_10k['large'])
larger = normalize(svdspace_10k['larger'])
low = normalize(svdspace_10k['low'])
lower = normalize(svdspace_10k['lower'])

find_similar_to(large - (low - lower), svdspace_10k)[:10]

[('large', 0.9106217738063563),
 ('small', 0.9005027565097975),
 ('larger', 0.8449501667244497),
 ('smaller', 0.8369244453991077),
 ('upper', 0.8027971582330167),
 ('into', 0.7818906203893848),
 ('along', 0.7729787978861417),
 ('within', 0.7687292227121255),
 ('remaining', 0.7650781899330422),
 ('vast', 0.759744201427435)]

In [15]:
#words similar to vast
big = normalize(svdspace_10k['big'])
small = normalize(svdspace_10k['small'])
vast = normalize(svdspace_10k['vast'])
tiny = normalize(svdspace_10k['tiny'])

find_similar_to(big - (tiny - small), svdspace_10k)[:10]

[('big', 0.8564092944467012),
 ('little', 0.767548853956461),
 ('large', 0.7540876021358716),
 ('small', 0.7535934995612573),
 ('like', 0.7391251045267836),
 ('well', 0.7266583736711634),
 ('while', 0.7233492341119214),
 ('just', 0.7117700176853392),
 ('records', 0.708139429859657),
 ('where', 0.7049662985566375)]

In [156]:
#words similar to cool
hot = normalize(svdspace_10k['hot'])
warm = normalize(svdspace_10k['warm'])
cool = normalize(svdspace_10k['cool'])
cold = normalize(svdspace_10k['cold'])

find_similar_to(hot - (cold - warm), svdspace_10k)[:10]

[('hot', 0.8859561826985469),
 ('warm', 0.8776172287973323),
 ('cool', 0.8361104173222172),
 ('wet', 0.8170370370501621),
 ('dry', 0.8014640389436251),
 ('honey', 0.7859380992894108),
 ('rough', 0.7671458833589136),
 ('flowers', 0.7625996175628357),
 ('soft', 0.7620242109742615),
 ('grass', 0.7539392936754656)]

In [113]:
#words similar to evening
day = normalize(svdspace_10k['day'])
morning = normalize(svdspace_10k['morning'])
night = normalize(svdspace_10k['night'])
evening = normalize(svdspace_10k['evening'])

find_similar_to(day - (morning - night), svdspace_10k)[:10]

[('day', 0.9079230332466901),
 ('place', 0.8494988502232454),
 ('night', 0.8364827163841919),
 ('before', 0.8339018112967359),
 ('while', 0.8287182968738214),
 ('last', 0.8047421061439832),
 ('after', 0.7959099696647033),
 ('just', 0.7723324496326847),
 ('during', 0.7705276128946046),
 ('next', 0.7684391173990448)]

In [115]:
#words similar to old
new = normalize(svdspace_10k['new'])
fresh = normalize(svdspace_10k['fresh'])
old = normalize(svdspace_10k['old'])
oldest = normalize(svdspace_10k['oldest'])

find_similar_to(new - (fresh - oldest), svdspace_10k)[:10]

[('oldest', 0.8030008653125098),
 ('new', 0.7731023936735686),
 ('historic', 0.7459930704484694),
 ('campus', 0.7018255746612554),
 ('metropolitan', 0.6733602238394403),
 ('municipal', 0.6669471208305824),
 ('museum', 0.6640370467049758),
 ('victorian', 0.6637230707146964),
 ('association', 0.6636921975110932),
 ('regional', 0.6620474128322882)]

## 5. Semantic composition and phrase similarity **[20 marks]**

In this task, we are going to look at how different semantic composition models, introduced in [2] correlate with human judgements. The file with the dataset is `mitchell_lapata_acl08.txt` included with this notebook. Your task is to do the following:  

(i) process the dataset, extract pairs of `reference - landmark high` and `reference - landmark low`; you can use the code from the lecture as something to start with. Note that there are 2 landmarks for each reference: one landmark exhibits high similarity with the reference, while another one has low similarity with the reference. A single human participant could have evaluated both of these pairs. For more details, we refer you to the paper.  

(ii) build models of semantic phrase composition: in the lecture we introduced simple additive, simple multiplicative and combined models (details are in [2]). Your task is to take a single pair (a reference or a high similarity landmark or a low similarity landmark) and compute the composition of its vectors using each of these functions. Thus, you will have three compositional models that take a `noun - verb` pair and output a single vector, representing the meaning of this pair. As your semantic space, you can use pretrained spaces (standard space, ppmi or svd) introduced above. It is up to you which space you use, but for someone who runs your code, it should be pretty straightforward to switch between them.

(iii) calculate Spearman correlation between each model's predictions and human judgements; you should have something similar to the scores that are shown in the paper [2]:  

![title](./res.png)

The paper states that they calculated correlations between each individual participant's judgeements and each model's predictions.  

Let's say we have 3 models: simple additive (A), simple multiplicative (M), combined (C).  
From our task dataset, we also know that we have 20 participants.  
Now, for each participant in 20 participants we get all `verb - noun` pairs that these participated evaluated.  
For example:

In [118]:
participant_judgemenets_example = [
 'participant50 chatter child gabble 6 high',
 'participant50 chatter tooth click 2 high',
 'participant50 reel head whirl 5 high',
 'participant50 reel mind stagger 4 low',
 'participant50 reel industry stagger 5 high',
 'participant50 reel man whirl 3 low',
 'participant50 glow fire beam 7 low',
 'participant50 glow face burn 3 low',
 'participant50 glow cigar burn 5 high',
 'participant50 glow skin beam 7 high'
    
]

In [119]:
participant_judgemenets_example

['participant50 chatter child gabble 6 high',
 'participant50 chatter tooth click 2 high',
 'participant50 reel head whirl 5 high',
 'participant50 reel mind stagger 4 low',
 'participant50 reel industry stagger 5 high',
 'participant50 reel man whirl 3 low',
 'participant50 glow fire beam 7 low',
 'participant50 glow face burn 3 low',
 'participant50 glow cigar burn 5 high',
 'participant50 glow skin beam 7 high']

Let's look at the first pair that participant50 evaluated: reference `child chatter` and high-level similarity landmark (as the last word in the row indicates) `child gabble`. The human gave the similarity score of 6 (very similar). Thus, human similarity judgment = [6].  

Our A model's output:  
cosine(p1, p2) = 0.88, where p1 is the result of addition of word vectors in the reference phrase `child gabble`, and p2 is the result of addition of word vectors in the high-level similarity phrase `child chatter`.  

Therefore, we have human rating vector [6] and model A output [0.88]. Next is to compute correlation between these two vectors.

To get an overall score, simply average your correlation scores over all participants, since you are calculating correlation scores per participant.

Of course, your human rating vectors will be longer (e.g., [6, 7, 3, 4, 5]) where each element is a participant's judgement of a specific pair. Each of your models (A, B, C) will produce a single vector of cosine similarity between these same pairs (e.g., [0.89, 0.98, 0.23, 0.65, 0.55]). The goal is to compare each model's cosine similarity vectors with human rating vectors and identify the model which outputs the best result in terms of being the closest to the way human rate similarity between the phrases.

The minimum to do in this task: compute correlations for 3 models mentioned above and human rating for AT LEAST one participant. Elaborate on how different the resulting correlation scores are depending on the model's composition function (additive, multiplicative, combined). For examples on how to interpret the results, look at Section 5 Results of the original paper.

In [150]:
#your code should go here...
dataset = {}
references = []
with open('mitchell_lapata_acl08.txt', 'r') as f:
    phrase_dataset = f.read().splitlines()
for line in phrase_dataset[1:]:
    participant_id, reference, noun, landmark, rating, hilo = line.split()
    reference_phrase = [noun, reference]
    landmark_phrase = [noun, landmark]
    if reference_phrase not in references:
        references.append(reference_phrase)
    if participant_id not in dataset:
        dataset[participant_id] = []
    else:
        dataset[participant_id].append((reference_phrase, landmark_phrase, rating, hilo))
dataset['participant20']  


{'participant20': [(['discussion', 'stray'], ['discussion', 'digress'], '6', 'high'), (['eye', 'stray'], ['eye', 'roam'], '7', 'high'), (['child', 'stray'], ['child', 'digress'], '1', 'low'), (['body', 'throb'], ['body', 'pulse'], '5', 'high'), (['head', 'throb'], ['head', 'shudder'], '2', 'low'), (['voice', 'throb'], ['voice', 'shudder'], '3', 'low'), (['vein', 'throb'], ['vein', 'pulse'], '6', 'high'), (['machine', 'chatter'], ['machine', 'click'], '4', 'high'), (['girl', 'chatter'], ['girl', 'gabble'], '6', 'high'), (['child', 'chatter'], ['child', 'click'], '4', 'low'), (['tooth', 'chatter'], ['tooth', 'gabble'], '3', 'low'), (['shot', 'rebound'], ['shot', 'ricochet'], '7', 'high'), (['ball', 'rebound'], ['ball', 'rally'], '2', 'low'), (['share', 'rebound'], ['share', 'ricochet'], '6', 'low'), (['optimism', 'rebound'], ['optimism', 'rally'], '4', 'high'), (['tongue', 'flicker'], ['tongue', 'waver'], '4', 'low'), (['screen', 'flicker'], ['screen', 'flick'], '2', 'high'), (['interest

**Any comments/thoughts should go here:**

# Literature

  - [1] C. Silberer and M. Lapata. Learning grounded meaning representations with autoencoders. In Proceedings of the 52nd Annual Meeting of the Association for Computational Linguistics, pages 721–732, Baltimore, Maryland, USA, June 23–25 2014 2014. Association for Computational Linguistics.  

  - [2] Mitchell, J., & Lapata, M. (2008). Vector-based Models of Semantic Composition. In Proceedings of ACL-08: HLT (pp. 236–244). Association for Computational Linguistics.
  
  - [3] T. Mikolov, I. Sutskever, K. Chen, G. S. Corrado, and J. Dean. Distributed representations of words and phrases and their compositionality. In Advances in neural information processing systems, pages 3111–3119, 2013.

## Marks

This assignment has a total of 60 marks.